shared/utils/typedefs-string.hpp

```diff
@@ -102,19 +117,110 @@ struct StringW {
+        return inst == nullptr;
+    }
+    #pragma clang diagnostic pop
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    StringW& operator +=(const T rhs) noexcept {
+        inst = strappend(inst, rhs);
+        return *this;
+    }
+
+    StringW& operator +=(StringW rhs) noexcept {
+        inst = il2cpp_utils::detail::alloc_str(this->operator std::u16string().append(rhs));
+        return *this;
+    }
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    StringW operator +(const T rhs) const noexcept {
+        return strappend(inst, rhs);
+    }
+
+    StringW operator +(StringW rhs) const noexcept {
+        return this->operator std::u16string().append(rhs);
+    }
+
+    bool operator <(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    bool operator <(const T rhs) const noexcept {
+        return strless(inst, rhs);
+    }
+
+    template<int sz>
+    constexpr bool operator ==(ConstString<sz> const& rhs) const noexcept {
+        Il2CppChar* first = inst->chars;
+        char16_t const* second = rhs.chars;
+        Il2CppChar* firstEnd = first + inst->length;
+        char16_t const* secondEnd = second + sz;
+
+        while (first != firstEnd && second != secondEnd)
+        {
+            if (*first != *second) return false;
+            first++; second++;
+        }
+
+        return first == firstEnd && second == secondEnd;
+    }
+    
+    bool operator ==(StringW rhs) const noexcept;
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    bool operator ==(const T rhs) const noexcept {
+        return streq(inst, rhs);
+    }
+
+    bool starts_with(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    bool starts_with(const T rhs) const noexcept {
+        return strstart(inst, rhs);
+    }
+    
+    bool ends_with(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
+    bool ends_with(const T rhs) const noexcept {
+        return strend(inst, rhs);
+    }
+    
+    // who needs bounds checking amiright
+    constexpr const Il2CppChar& operator [](size_t idx) const { return inst->chars[idx]; }
```

Comment:
Very reasonable, though easy const! and also, it might make sense to actually expose a non-const indexing operator for cool people to modify strings in place without having to make new strings and copy, etc. Seems like a simple-enough add, so I'd recommend it.